### PR TITLE
AIP-84 Migrate GET Dag Run endpoint to FastAPI

### PR DESCRIPTION
**Pull request by rawwar**
related to: #42701

### CODE REVIEW

File: airflow/api_fastapi/serializers/dag_run.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    start_date: datetime | None
+    end_date: datetime | None
+    data_interval_start: datetime | None
+    data_interval_end: datetime | None
+    last_scheduling_decision: datetime | None
+    run_type: DagRunType
+    state: DagRunState
+    external_trigger: bool
+    triggered_by: DagRunTriggeredByType
+    conf: dict
```
Comment (rawwar) [inline airflow/api_fastapi/serializers/dag_run.py:43]: @pierrejeambrun , Should I be including `note` field here?  If so, Should I be querying `dag_run_note` table and return it as a json text?

Comment (bbovenzi) [reply airflow/api_fastapi/serializers/dag_run.py:43]: I believe so. A user shouldn't need to send a second request just for the note text.

Comment (pierrejeambrun) [reply airflow/api_fastapi/serializers/dag_run.py:43]: +1. The legacy endpoint returns the note, so the new endpoint should also return it.

It's a special 'easy' case because I believe there's already a relationshiop and an association_proxy on the `content` from `dag_run_note` table. All that to say that we do not even need to manually join anything, `.note` should be accessible.

Maybe we can just pre-load the relationship to make the serialization step faster. (and not emit many request caused by the lazy loading), we can achieve that In the select query through the `.options` method.

Comment (rawwar) [reply airflow/api_fastapi/serializers/dag_run.py:43]: Thanks. I am updating the code

Comment (rawwar) [reply airflow/api_fastapi/serializers/dag_run.py:43]: I just started looking into this and realised, dag_maker cannot add a note to dag_run. Even in the earlier endpoint, during POST call to create Dag run has a separate condition which checks presence of `note` and adds it using `association_proxy`. 

So, I should be first adding dag_run_note creation functionality to `dag_maker` and then proceed to write a test. I should also keep in mind to add dag_run_note when migrating POST create_dag_run endpoint. 

Or I can just test note=None in response. This seems easier

Comment (pierrejeambrun) [reply airflow/api_fastapi/serializers/dag_run.py:43]: For the `POST` I need we need to be iso-functional with the legacy endpoint. If it supports `POST` with a `note` then the new endpoint should too. If not then it's fine to not having it in the first iteration :) (I know there's a couple of dedicated endpoints to set a note)

Comment (rawwar) [reply airflow/api_fastapi/serializers/dag_run.py:43]: For now, I updated tests to just check if `note is None`. Once I add endpoint to trigger dag run, I'll update tests to include testing note with something other than None. 

I also added a linked PR: https://github.com/apache/airflow/pull/42800 where I added  DagRunNotePydantic class.

Comment (pierrejeambrun) [reply airflow/api_fastapi/serializers/dag_run.py:43]: You can just add manually a note to any of the dag model, to be able to test not `None` notes in this PR.

You can take a look at to see how you can create a Note on a dagrun:
```
@security.requires_access_dag("PUT", DagAccessEntity.RUN)
@action_logging
@provide_session
def set_dag_run_note(*, dag_id: str, dag_run_id: str, session: Session = NEW_SESSION) -> APIResponse:
    """Set the note for a dag run."""
    dag_run: DagRun | None = session.scalar(
        select(DagRun).where(DagRun.dag_id == dag_id, DagRun.run_id == dag_run_id)
    )
    if dag_run is None:
        error_message = f"Dag Run id {dag_run_id} not found in dag {dag_id}"
        raise NotFound(error_message)
    try:
        post_body = set_dagrun_note_form_schema.load(get_json_request_dict())
        new_note = post_body["note"]
    except ValidationError as err:
        raise BadRequest(detail=str(err))

    current_user_id = get_auth_manager().get_user_id()
    if dag_run.dag_run_note is None:
        dag_run.note = (new_note, current_user_id)
    else:
        dag_run.dag_run_note.content = new_note
        dag_run.dag_run_note.user_id = current_user_id
    session.commit()
    return dagrun_schema.dump(dag_run)

```

Basically those line for `update or create` a dag note:
```python
current_user_id = get_auth_manager().get_user_id()
    if dag_run.dag_run_note is None:
        dag_run.note = (new_note, current_user_id)
    else:
        dag_run.dag_run_note.content = new_note
        dag_run.dag_run_note.user_id = current_user_id
```

Comment (rawwar) [reply airflow/api_fastapi/serializers/dag_run.py:43]: All I had to do was add a note as type `str`, and sqlalchemy did its magic!  

Thanks @pierrejeambrun . I updated the PR with the test for dag_note(non-None)

**Comment by rawwar** [comment]
@pierrejeambrun , @bbovenzi , Can. you please add "legacy api" label to this PR

File: airflow/api_fastapi/views/public/dag_run.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+from sqlalchemy.orm import Session
+from typing_extensions import Annotated
+
+from airflow.api_fastapi.db.common import get_session
+from airflow.api_fastapi.openapi.exceptions import create_openapi_http_exception_doc
+from airflow.api_fastapi.serializers.dag_run import DAGRunResponse
+from airflow.api_fastapi.views.router import AirflowRouter
+from airflow.models import DagRun
+
+dag_run_router = AirflowRouter(tags=["DagRun"])
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/views/public/dag_run.py]: To avoid having to copy/past this part on all `@router`

```suggestion
dag_run_router = AirflowRouter(tags=["DagRun"], prefix="/dags/{dag_id}/dagRuns")
```

File: tests/api_fastapi/views/public/test_dag_run.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+@pytest.mark.parametrize(
+    "dag_id, run_id, state, run_type, triggered_by, dag_run_note",
+    [
+        (DAG1_ID, DAG1_RUN1_ID, DAG1_RUN1_STATE, DAG1_RUN1_RUN_TYPE, DAG1_RUN1_TRIGGERED_BY, DAG1_NOTE),
+        (DAG1_ID, DAG1_RUN2_ID, DAG1_RUN2_STATE, DAG1_RUN2_RUN_TYPE, DAG1_RUN2_TRIGGERED_BY, None),
+        (DAG2_ID, DAG2_RUN1_ID, DAG2_RUN1_STATE, DAG2_RUN1_RUN_TYPE, DAG2_RUN1_TRIGGERED_BY, None),
+        (DAG2_ID, DAG2_RUN2_ID, DAG2_RUN2_STATE, DAG2_RUN2_RUN_TYPE, DAG2_RUN2_TRIGGERED_BY, None),
+    ],
+)
+def test_get_dag_run(test_client, dag_id, run_id, state, run_type, triggered_by, dag_run_note):
```
Comment (pierrejeambrun) [inline tests/api_fastapi/views/public/test_dag_run.py:121]: Maybe opt for `class oriented` tests. For instance you can take a look to `connections` tests.

I will update `dags`, because this is more consistant with `class oriented` airflow tests. (most of them are written like this).

Also you can add a test case for 404.

**Review by pierrejeambrun — COMMENTED** [review]
Looking good.

A few minor suggestions, then we can merge :). Needs rebase re-rerun code gen

File: airflow/api_fastapi/views/public/dag_run.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+from airflow.api_fastapi.openapi.exceptions import create_openapi_http_exception_doc
+from airflow.api_fastapi.serializers.dag_run import DAGRunResponse
+from airflow.api_fastapi.views.router import AirflowRouter
+from airflow.models import DagRun
+
+dag_run_router = AirflowRouter(tags=["DagRun"], prefix="/dags/{dag_id}/dagRuns")
+
+
+@dag_run_router.get(
+    "/dags/{dag_id}/dagRuns/{dag_run_id}", responses=create_openapi_http_exception_doc([401, 403, 404])
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/views/public/dag_run.py]:
```suggestion
    "/{dag_run_id}", responses=create_openapi_http_exception_doc([401, 403, 404])
```

Comment (pierrejeambrun) [reply airflow/api_fastapi/views/public/dag_run.py]: Then you can remove it from the path because you have it defined in the `prefix` of the router I think. Not sure how path param will behave but I think it works.

**Review by pierrejeambrun — COMMENTED** [review]
Let me know when you added the 404 tests. Looking good overall, minor suggestion.

File: airflow/api_fastapi/views/public/dag_run.py (hunk 4 above)
Comment (rawwar) [reply airflow/api_fastapi/views/public/dag_run.py]: Thanks. I'll test this

**Comment by rawwar** [comment]
> Let me know when you added the 404 tests. Looking good overall, minor suggestion.

Do you want me to update tests to Class based on this PR or another one?

File: tests/api_fastapi/views/public/test_dag_run.py (hunk 3 above)
Comment (rawwar) [reply tests/api_fastapi/views/public/test_dag_run.py:121]: @pierrejeambrun , Do you want me to update tests to Class based on this PR or another one?

**Comment by rawwar** [comment]
> Let me know when you added the 404 tests. Looking good overall, minor suggestion.

I've added a test for 404.

File: tests/api_fastapi/views/public/test_dag_run.py (hunk 3 above)
Comment (rawwar) [reply tests/api_fastapi/views/public/test_dag_run.py:121]: I would like to work on updating tests to class-based tests for both tests_dag.py and tests_dag_run.py in a separate PR. Hope that's fine?

Created an issue- https://github.com/apache/airflow/issues/42866#issue-2576446788

Comment (pierrejeambrun) [reply tests/api_fastapi/views/public/test_dag_run.py:121]: Ok, assigned :)

File: airflow/api_fastapi/serializers/dag_run.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/serializers/dag_run.py:43]: Compared to the legacy endpoint, I think we are missing `execution_date` is that intended ? Both logical_date and execution_date seem to be in the code and in the spec:

https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/get_dag_run

Comment (rawwar) [reply airflow/api_fastapi/serializers/dag_run.py:43]: I thought, since its deprecated, We are ok to do a breaking change for Airflow 3.x. Should I add it?

Comment (pierrejeambrun) [reply airflow/api_fastapi/serializers/dag_run.py:43]: You are right, lets remove it right away, this will get deleted in airflow 3.

**Review by pierrejeambrun — APPROVED** [review]
